package/gluon-ebtables-limit-arp/src/gluon-arp-limiter.c

```diff
@@ -0,0 +1,306 @@
+{
+	char str[128];
+	int ret;
+
+	snprintf(str, sizeof(str),
+		 EBTABLES " %s ARP_LIMIT_TLCHECK %s --source %s -j DROP",
+		 mod, add ? "2" : "", mac_ntoa(mac));
+
+	ret = system(str);
+	if (ret)
+		fprintf(stderr,
+			"%i: Calling ebtables for TL failed with status %i\n",
+			clock, ret);
+}
+
+static void ebt_mac_call(char *mod, struct mac_addr *mac)
+{
+	if (!strncmp(mod, "-D", strlen(mod))) {
+		ebt_mac_ret_call(mod, mac, 0);
+		ebt_mac_limit_call(mod, mac);
+	} else {
+		ebt_mac_limit_call(mod, mac);
+		ebt_mac_ret_call(mod, mac, 1);
+	}
+}
+
+static void mac_node_destructor(struct addr_list *node)
+{
+	struct mac_addr *mac = (struct mac_addr *)node->addr;
+
+	ebt_mac_call("-D", mac);
+}
+
+static int dat_parse_line(const char *line, struct in_addr *ip)
+{
+	int ret;
+	char *p;
+	char *tok;
+
+	p = strpbrk(line, "0123456789");
+	if (!p) {
+		fprintf(stderr, "Error: Can't find integer in: %s\n", line);
+		return -EINVAL;
+	}
+
+	tok = strtok(p, " ");
+	if (!tok) {
+		fprintf(stderr, "Error: Can't find end of string': %s\n", line);
+		return -EINVAL;
+	}
+
+	ret = inet_aton(p, ip);
+	if (!ret) {
+		fprintf(stderr, "Error: inet_aton failed on: %s\n", p);
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static void ebt_add_ip(struct in_addr ip)
+{
+	int ret = addr_store_add(&ip, &ip_store);
+
+	/* already stored or out-of-memory */
+	if (ret)
+		return;
+
+	ebt_ip_call("-I", ip);
+}
+
+static void ebt_add_mac(struct mac_addr *mac)
+{
+	int ret = addr_store_add(mac, &mac_store);
+
+	/* already stored or out-of-memory */
+	if (ret)
+		return;
+
+	ebt_mac_call("-I", mac);
+}
+
+static void ebt_dat_update(void)
+{
+	FILE *fp;
+	char line[256];
+	char *pline;
+	int ret;
+	struct in_addr ip;
+
+	fp = popen(BATCTL_DC, "r");
+	if (!fp) {
+		fprintf(stderr, "%i: Error: Could not call batctl dc\n", clock);
+		return;
+	}
+	
+	while (!feof(fp)) {
+		pline = fgets(line, sizeof(line), fp);
+		if (!pline) {
+			fprintf(stderr, "%i: Error: fgets() failed\n", clock);
+			break;
+		}
+
+		ret = dat_parse_line(line, &ip);
+		if (ret < 0) {
+			fprintf(stderr, "%i: Error: Parsing line failed\n",
+				clock);
+			break;
+		}
+
+		ebt_add_ip(ip);
+	}
+
+	pclose(fp);
+}
+
+static int tl_parse_line(char *line, struct mac_addr *mac)
+{
+	int ret;
+	char *p;
+	char *tok;
+
+	p = strpbrk(line, "0123456789abcdef");
+	if (!p) {
+		fprintf(stderr, "Error: Can't find hex in: %s\n", line);
+		return -EINVAL;
+	}
+
+	tok = strtok(p, " ");
+	if (!tok) {
+		fprintf(stderr, "Error: Can't find end of string': %s\n", line);
+		return -EINVAL;
+	}
+
+	ret = mac_aton(p, mac);
+	if (!ret) {
+		fprintf(stderr, "Error: mac_aton failed on: %s\n", p);
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static void ebt_tl_update(void)
+{
+	FILE *fp;
+	char line[256];
+	char *pline;
+	int ret;
+	struct mac_addr mac;
+
+	fp = popen(BATCTL_TL, "r");
+	if (!fp) {
+		fprintf(stderr, "%i: Error: Could not call batctl tl\n", clock);
+		return;
+	}
+
+	// Throw away a header line
```

Comment:
Why are you throwing a header line away when you called batctl without header lines (-H)?

Reply:
I see this output with batman-adv-legacy / batctl v2013.4:

<pre><code>root@freifunk647002d16222:~# batctl -v
batctl 2013.4.0 [batman-adv: 2013.4.0]
root@freifunk647002d16222:~# batctl tl
Locally retrieved addresses (from bat0) announced via TT (TTVN: 2 CRC: 0x17b1):
       Client        Flags   Last seen 
 * 64:70:02:d1:62:22 [.P...]   0.000
root@freifunk647002d16222:~# batctl tl -H
       Client        Flags   Last seen 
 * 64:70:02:d1:62:22 [.P...]   0.000
root@freifunk647002d16222:~#</code></pre>

So looks like a bug in that batctl version, hm. Could someone tell me the output they see with a more recent compat15 batman-adv/batctl?

Reply:
Oh, funny, that was actually a patch from me fixing it some years ago:

https://git.open-mesh.org/batctl.git/commit/4aa06a7c2935158a2ff5ffea2995b011a62e229e

I'll backport it for batctl-legacy.

PS: And yes, current batctl does need need that extra skipped line, just confirmed it on some device. Thanks for reporting!

Reply:
With 

    root@PL-See68-DG:~# batctl -v
    batctl 2017.2 [batman-adv: 2017.2]
    root@PL-See68-DG:~# batctl tl
    [B.A.T.M.A.N. adv 2017.2, MainIF/MAC: primary0/0e:79:de:08:8b:d3 (bat0/ac:86:74:00:ec:00     BATMAN_IV), TTVN: 215]
    Client             VID Flags    Last seen (CRC       )
    ac:86:74:00:ec:00   -1 [.P....]   0.000   (0x8ace0d4d)
    f4:09:d8:2b:51:78   -1 [....W.] 167.190   (0x8ace0d4d)
    ac:86:74:00:ec:00    0 [.P....]   0.000   (0x55a68955)
    root@PL-See68-DG:~# batctl tl -H
    ac:86:74:00:ec:00   -1 [.P....]   0.000   (0x8ace0d4d)
    f4:09:d8:2b:51:78   -1 [....W.] 170.620   (0x8ace0d4d)
    ac:86:74:00:ec:00    0 [.P....]   0.000   (0x55a68955)